### PR TITLE
Bug/issue 115/smart fields always calculated

### PR DESCRIPTION
**Pull request by lucasob**
This PR aims to address Issue #115 where smart fields would be filtered out of a response, but only after being evaluated. This put our API under great load, so rather than just calculate and then filter, this change will just ignore smart fields that weren't requested.

A [support ticked](https://community.forestadmin.com/t/dont-load-smart-field-in-collection-view/5102) was also raised.

### CODE REVIEW

**Comment by vamonte** [comment]
Thks @lucasob, I ll check your PR soon

**Comment by lucasob** [comment]
@vamonte Bumping this please :)

**Comment by vamonte** [comment]
I'm reviewing today

File: django_forest/resources/utils/smart_field.py [hunk 1]
```diff
@@ -24,13 +24,33 @@ def _add_smart_fields(self, item, smart_fields, resource):
         for smart_field in smart_fields:
             self._handle_get_method(smart_field, item, resource)
 
-    def handle_smart_fields(self, queryset, resource, many=False):
+    def _get_smart_fields_for_request(self, collection, params):
```
Comment (vamonte) [inline django_forest/resources/utils/smart_field.py:27]: I'm not a big fan of this method.
Should be better to use something like that:
````python
def _get_smart_fields_for_request(self, collection, params=None):
  fields = [field for field in collection['fields'] if field['is_virtual']]
  if params and 'fields' in params:
    try:
      queried_fields = params['fields'][collection['name']]
    except KeyError:
      # should be never raised
      pass
    else:
      fields = [field for field in fields if field in queried_fields]
  return fields
````

File: django_forest/resources/views/detail.py [hunk 2]
```diff
@@ -24,7 +24,7 @@ def get_instance(self, request, pk):
 
         instance = queryset[0]
         # handle smart fields
-        self.handle_smart_fields(instance, self.Model._meta.db_table)
+        self.handle_smart_fields(instance, self.Model._meta.db_table, None)
```
Comment (vamonte) [inline django_forest/resources/views/detail.py:27]: __params__ should be an optional argument

File: django_forest/resources/views/detail.py [hunk 3]
```diff
@@ -35,7 +35,7 @@ def get(self, request, pk):
             return self.error_response(e)
         else:
             # handle smart fields
-            self.handle_smart_fields(instance, self.Model._meta.db_table)
+            self.handle_smart_fields(instance, self.Model._meta.db_table, None)
```
Comment (vamonte) [inline django_forest/resources/views/detail.py:38]: __params__ should be an optional argument

**Review by vamonte — CHANGES_REQUESTED** [review]
@lucasob 
Thank's for the report and the PR, we need to add this fix.
Could you fix your PR and I'll merge it.

File: django_forest/resources/utils/smart_field.py (hunk 1 above)
Comment (lucasob) [reply django_forest/resources/utils/smart_field.py:27]: Hey, I feel a little uncomfortable with this solution, as it seems a little less readable. I'll try and meet in the middle, but ultimately that much nesting with so many conditions/catches combined with the mutation doesn't make for clear logic in my honest opinion.

Comment (vamonte) [reply django_forest/resources/utils/smart_field.py]: I'm not sure about your new version.  
I'm going to investigate more about this issue to be sure on the new implem.

Comment (lucasob) [reply django_forest/resources/utils/smart_field.py]: Ok, happy to work with you on this. This issue is capable of halting production so we will have to run off a fork until this is merged.

**Review by vamonte — APPROVED** [review]

**Comment by vamonte** [comment]
@lucasob flake8 seems broken yet.

**Comment by lucasob** [comment]
Hey @vamonte can we get CI running again please? Thanks!

**Comment by lucasob** [comment]
commitlint should now be happy! Thank you @vamonte !

**Comment by vamonte** [comment]
@lucasob I'll merge it today

**Pull request merged**